src/_pytest/main.py

```diff
@@ -958,16 +964,41 @@ def collect(self) -> Iterator[nodes.Item | nodes.Collector]:
             self.trace.root.indent -= 1
 
     def genitems(self, node: nodes.Item | nodes.Collector) -> Iterator[nodes.Item]:
+        import inspect
+
+        from _pytest.python import Class
+        from _pytest.python import Function
+        from _pytest.python import Module
+
         self.trace("genitems", node)
         if isinstance(node, nodes.Item):
             node.ihook.pytest_itemcollected(item=node)
+            if not self.config.getini("collect_imported_tests"):
+                if isinstance(node.parent, Module) and isinstance(node, Function):
+                    if inspect.isfunction(node._getobj()):
+                        fn_defined_at = node._getobj().__module__
+                        in_module = node.parent._getobj().__name__
+                        if fn_defined_at != in_module:
+                            return
             yield node
         else:
             assert isinstance(node, nodes.Collector)
             keepduplicates = self.config.getoption("keepduplicates")
             # For backward compat, dedup only applies to files.
             handle_dupes = not (keepduplicates and isinstance(node, nodes.File))
             rep, duplicate = self._collect_one_node(node, handle_dupes)
+
+            if not self.config.getini("collect_imported_tests"):
```

Comment:
Same comment here, this logic should be moved to `_pytest.python` instead of `main`.